arrow-libs/core/arrow-core/src/commonMain/kotlin/arrow/core/Option.kt

```diff
@@ -812,19 +828,149 @@ fun <A> Iterable<Option<A>>.combineAll(MA: Monoid<A>): Option<A> =
     acc.combine(MA, a)
   }
 
-fun <T> Iterable<T>.firstOrNone(): Option<T> = this.firstOrNull().toOption()
+/**
+ * Returns the first element as [Some(element)][Some], or [None] if the iterable is empty.
+ */
+fun <T> Iterable<T>.firstOrNone(): Option<T> =
+  when (this) {
+    is Collection -> firstOrNone()
+    else -> iterator().nextOrNone()
+  }
+
+/**
+ * Returns the first element as [Some(element)][Some], or [None] if the collection is empty.
+ */
+fun <T> Collection<T>.firstOrNone(): Option<T> =
+  if (isEmpty()) {
+    None
+  } else {
+    Some(first())
+  }
+
+private fun <T> Iterator<T>.nextOrNone(): Option<T> =
+  if (hasNext()) {
+    Some(next())
+  } else {
+    None
+  }
+
+/**
+ * Returns the first element as [Some(element)][Some] matching the given [predicate], or [None] if element was not found.
+ */
+inline fun <T> Iterable<T>.firstOrNone(predicate: (T) -> Boolean): Option<T> {
+  for (element in this) {
+    if (predicate(element)) {
+      return Some(element)
+    }
+  }
+  return None
+}
+
+/**
+ * Returns single element as [Some(element)][Some], or [None] if the iterable is empty or has more than one element.
+ */
+fun <T> Iterable<T>.singleOrNone(): Option<T> =
+  when (this) {
+    is Collection -> singleOrNone()
+    else -> iterator().run { nextOrNone().filter { !hasNext() } }
+  }
+
+/**
+ * Returns single element as [Some(element)][Some], or [None] if the collection is empty or has more than one element.
+ */
+fun <T> Collection<T>.singleOrNone(): Option<T> =
```

Comment:
I suggested the smaller functions to avoid functions being too complex but guess it may make sense to have a single one of both show up for collections on autocomplete.